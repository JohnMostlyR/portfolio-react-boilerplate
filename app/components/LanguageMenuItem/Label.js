import styled from 'styled-components';
import PropTypes from 'prop-types';

const Label = styled.span`
  display: inline-block;
  margin-left: 1rem;
  font-size: 1rem;
  white-space: nowrap;

  &::after {
    content: ${({ isSelected }) => (isSelected ? '"\u2713"' : '""')};
    margin-left: 0.5em;
  }
`;

Label.propTypes = {
  isSelected: PropTypes.bool,
};

export default Label;
