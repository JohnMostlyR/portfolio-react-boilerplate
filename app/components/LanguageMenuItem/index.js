/**
 *
 * LanguageMenuItem
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import IconEN from './IconEN';
import IconNL from './IconNL';
import Label from './Label';

export const AVAILABLE_ICONS = ['en', 'nl'];

const Button = styled.button.attrs({
  type: 'button',
})`
  align-items: center;
  display: flex;
  padding: 1rem;
  width: 100%;

  &:hover {
    cursor: pointer;
  }
`;

function LanguageMenuItem({ value, onClickHandler, isSelected }) {
  function handleClick(lang, ev) {
    ev.preventDefault();
    onClickHandler(lang);
  }

  function handleKeyDown(lang, ev) {
    if (ev.keyCode === 13) {
      // Enter
      ev.preventDefault();
      onClickHandler(lang);
    }
  }

  const labels = {
    nl: (
      <Button
        role="menuitemradio"
        aria-checked={isSelected}
        disabled={isSelected}
        tabIndex="-1"
        lang="nl"
        data-hasfocus="false"
        onClick={el => handleClick('nl', el)}
        onKeyDown={el => handleKeyDown('nl', el)}
      >
        <IconNL isSelected={isSelected} />
        <Label isSelected={isSelected}>Nederlands</Label>
      </Button>
    ),
    en: (
      <Button
        role="menuitemradio"
        aria-checked={isSelected}
        disabled={isSelected}
        tabIndex="-1"
        lang="en"
        data-hasfocus="false"
        onClick={el => handleClick('en', el)}
        onKeyDown={el => handleKeyDown('en', el)}
      >
        <IconEN isSelected={isSelected} />
        <Label isSelected={isSelected}>English</Label>
      </Button>
    ),
  };

  return labels[value] || <React.Fragment />;
}

LanguageMenuItem.propTypes = {
  isSelected: PropTypes.bool,
  onClickHandler: PropTypes.func.isRequired,
  value: PropTypes.oneOf(AVAILABLE_ICONS).isRequired,
};

LanguageMenuItem.defaultProps = {
  isSelected: false,
};

export default LanguageMenuItem;
