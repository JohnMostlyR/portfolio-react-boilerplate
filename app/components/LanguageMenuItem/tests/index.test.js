import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import LanguageMenuItem from '../index';

describe('<LanguageMenuItem />', () => {
  const onClickHandler = jest.fn();

  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(
      <LanguageMenuItem value="nl" onClickHandler={onClickHandler} />,
    );
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "onClickHandler" property', () => {
    const preventDefault = jest.fn();
    const renderedComponent = shallow(
      <LanguageMenuItem value="en" onClickHandler={onClickHandler} />,
    );

    renderedComponent.simulate('Click', { preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(onClickHandler).toHaveBeenCalledTimes(1);
    expect(onClickHandler).toHaveBeenCalledWith('en');

    jest.clearAllMocks();

    renderedComponent.simulate('KeyDown', { preventDefault, keyCode: 13 });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(onClickHandler).toHaveBeenCalledTimes(1);
    expect(onClickHandler).toHaveBeenCalledWith('en');
  });

  it('should adopt the "isSelected" property', () => {
    const PROPERTY_VALUE = true;
    const renderedComponent = shallow(
      <LanguageMenuItem
        value="nl"
        onClickHandler={onClickHandler}
        isSelected={PROPERTY_VALUE}
      />,
    );

    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});
