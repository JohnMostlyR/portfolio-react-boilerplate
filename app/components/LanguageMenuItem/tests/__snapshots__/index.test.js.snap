// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageMenuItem /> should adopt the "isSelected" property 1`] = `
<LanguageMenuItem__Button
  aria-checked={true}
  data-hasfocus="false"
  disabled={true}
  lang="nl"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitemradio"
  tabIndex="-1"
>
  <IconNL
    isSelected={true}
  />
  <Label
    isSelected={true}
  >
    Nederlands
  </Label>
</LanguageMenuItem__Button>
`;

exports[`<LanguageMenuItem /> should render and match the snapshot 1`] = `
<LanguageMenuItem__Button
  aria-checked={false}
  data-hasfocus="false"
  disabled={false}
  lang="nl"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitemradio"
  tabIndex="-1"
>
  <IconNL
    isSelected={false}
  />
  <Label
    isSelected={false}
  >
    Nederlands
  </Label>
</LanguageMenuItem__Button>
`;
